### PR TITLE
llvm: Refactor 'is_finished' compilation

### PR DESCRIPTION
**Pull request by jvesely**
Add support for minimal history of stateful parameters.
Use function call instead of just reading the flags.
Consolidate restart of execution counters.
Restore O2 optimization level.
Add tests.

Closes: #1525

### CODE REVIEW

**Comment by jvesely** [comment]
@kmantel, f52693e adds check for history length in 'is_finished' computation. let me know if it looks ok.

**Comment by jvesely** [comment]
Restoring O2 partially fixes #1525.
Running the predator rey pytest now takes 1.5 mins instead of 40s, but that's much better than the originally reported 6 min.

File: psyneulink/core/llvm/helpers.py [hunk 1]
```diff
@@ -82,11 +82,32 @@ def get_param_ptr(builder, component, params_ptr, param_name):
...
+            "History not available: {} ({})".format(ptr.type.pointee, hist_idx)
+        ptr = builder.gep(state_ptr, [ir.IntType(32)(0), idx,
+                                      ir.IntType(32)(hist_idx)],
+                          name="ptr_state_{}_{}_hist{}".format(stateful_name,
+                                                               component.name,
+                                                               hist_idx))
+    return ptr
+
+
+def push_state_val(builder, component, state_ptr, name, new_val):
```
[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/helpers.py:102]: it abstract way the need to know the history length.
All stateful params that are not of `Component` type have a compiled history of at least 1 (see `componenet.py:1272`).

File: psyneulink/core/llvm/codegen.py [hunk 2]
```diff
@@ -75,7 +76,7 @@ def gen_node_wrapper(ctx, composition, node, *, tags:frozenset):
         node_in = builder.alloca(node_function.args[2].type.pointee)
         incoming_projections = node.afferents
 
-    if "reset" in tags:
+    if "reset" in tags or "is_finished" in tags:
```
[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/codegen.py:79]: I'm not sure what you mean by 'struct value'. We need to invoke 'is_finished' to match python behaviour. In the most complicated case, it means taking the last two 'value' ehm, values. and running the 'termination_measure' function

File: psyneulink/core/llvm/helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SamKG) [reply psyneulink/core/llvm/helpers.py:102]: there is only ever a `min_length` specified - is there also an upper limit?

[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/helpers.py:102]: there's both `history_min_length` and `history_max_length`. The semantics are that `min_length` is required for computation, max length is for users.
We can switch to using `max_length` when there's a convenient mechanism to give users access to those values.

[the remark this answers is not in the document]
Comment (jvesely) [reply psyneulink/core/llvm/helpers.py:102]: I'd be a bit worried about the impact on memory footprint and locality. We currently have the advantage of putting all data in one contiguous location that fits in the cache.

**Comment by kmantel** [comment]
Sorry I didn't notice the tag until now. I think 
```diff
diff --git a/psyneulink/core/components/mechanisms/processing/transfermechanism.py b/psyneulink/core/components/mechanisms/processing/tran
index 9bb3320be6..2fc5cb2064 100644
--- a/psyneulink/core/components/mechanisms/processing/transfermechanism.py
+++ b/psyneulink/core/components/mechanisms/processing/transfermechanism.py
@@ -1790,7 +1790,10 @@ class TransferMechanism(ProcessingMechanism_Base):
                    assert False, f"PROGRAM ERROR: Unable to determine length of input for" \
                                f" {repr(TERMINATION_MEASURE)} arg of {self.name}"

-        self.parameters.value.history_min_length = self._termination_measure_num_items_expected - 1
+        self.parameters.value.history_min_length = max(
+            self.parameters.value.history_min_length,
+            self._termination_measure_num_items_expected - 1
+        )

    def _report_mechanism_execution(self, input, params, output, context=None):
        """Override super to report previous_input rather than input, and selected params
```

could be better in case some time we figure out transfermechanism or a child class needs an even higher `history_min_length`. Although that's unlikely anyway, so not a huge deal.

**Pull request merged**